lxd/network/openvswitch/docs/developer-documentation.md

```diff
@@ -0,0 +1,127 @@
+
+This document outlines LXD's approach to OVN in a basic setup.
+It does not yet cover load-balancers, peering, forwards, zones, or ACLs.
+
+For more detailed documentation on OVN itself, please see:
+1. [Overview of OVN and SDNs](https://ubuntu.com/blog/data-centre-networking-what-is-ovn).
+2. [OVN architectural overview](https://manpages.ubuntu.com/manpages/jammy/man7/ovn-architecture.7.html).
+3. [OVN northbound database schema documentation](https://manpages.ubuntu.com/manpages/jammy/en/man5/ovn-nb.5.html).
+4. [OVN southbound database schema documentation](https://manpages.ubuntu.com/manpages/jammy/en/man5/ovn-sb.5.html).
+
+## OVN concepts
+This section outlines the OVN concepts that we use in LXD.
+These are usually represented in tables in the OVN northbound database.
+
+### Chassis
+A chassis is where traffic physically ingresses to or egresses from the virtual network.
+In LXD, there will usually be one chassis per cluster member.
+If LXD is configured to use OVN networking, then all members *can* be used as OVN chassis.
+
+> If any cluster members have the role `ovn-chassis`, only those members will be represented as chassis in the chassis group table (see below).
+> If no members have the role, all cluster members will be added to the chassis group.
+
+### OVS (OpenVSwitch) Bridge
+OVS bridges are used to connect virtual networks to physical ones and vice-versa.
+In LXD, if OpenVSwitch APIs are invoked directly, that means changes are being applied on the same physical cluster member.
+
+For each LXD cluster member there are two OVS bridges:
+1. The uplink bridge - this connects the uplink network on the host to the external switch inside each OVN network.
+2. The integration bridge - this connects instances to the internal switch inside each OVN network.
+
+### Chassis group
+A chassis group is an indirection between physical chassis and the virtual networks that use them.
+Each LXD OVN network has one chassis group.
+This allows us to, for example, set chassis priority on a per-network basis so that not all ingress/egress is happening on a single cluster member.
+
+### OVN underlay
+The OVN underlay is the means by which networks are virtualized across cluster members.
+It is a Geneve tunnel which creates a layer 2 overlay network across layer 3 infrastructure.
+The OVN underlay is configured and managed by OVN.
+
+### Logical router
+A logical router is a virtualized router.
+There is one per LXD OVN network.
+This handles L3 networking and additionally has associated NAT rules and security policies.
+
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
```

Comment:
"the to uplink"?

Reply:
Hmm yeah I meant uplink bridge here to mean the OVS bridge that sits before the uplink network.